src/TreeRegression.cpp

```diff
@@ -862,25 +915,54 @@ void TreeRegression::findBestSplitValueBeta(size_t nodeID, size_t varID, double
   }
 }
 
-void TreeRegression::addImpurityImportance(size_t nodeID, size_t varID, double decrease) {
+void TreeRegression::addImpurityImportance(size_t nodeID, size_t best_varID, double decrease,
+                                           uint use_depth,
+                                           std::vector<double> coef_reg,
+                                           int depth) {
 
   size_t num_samples_node = end_pos[nodeID] - start_pos[nodeID];
+  double diff; 
+  int next_depth; 
+  next_depth = depth + 1;
+  double best_decrease = decrease; 
+  
 
-  double best_decrease = decrease;
   if (splitrule != MAXSTAT) {
     double sum_node = 0;
     for (size_t pos = start_pos[nodeID]; pos < end_pos[nodeID]; ++pos) {
       size_t sampleID = sampleIDs[pos];
       sum_node += data->get_y(sampleID, 0);
     }
-    best_decrease = decrease - sum_node * sum_node / (double) num_samples_node;
+    
+    // accounting for the regularization
+    if((*all_split_varIDs)[best_varID] == 1){
+      diff = (sum_node * sum_node / (double) num_samples_node);
+      best_decrease = decrease - diff; 
+    } else{  
+      if(use_depth == 1){
+        diff = ((sum_node * sum_node / (double) num_samples_node) * std::pow(coef_reg[best_varID], next_depth));
+        best_decrease = decrease - diff; 
+      } else {
+        diff = ((sum_node * sum_node / (double) num_samples_node) * coef_reg[best_varID]);
+        best_decrease = decrease - diff; 
+      }
+    }
+
+    if ((*all_split_varIDs)[best_varID] != 1){
+      if (use_depth == 1){  
+        best_decrease = best_decrease * std::pow(coef_reg[best_varID], next_depth);
+      } else {
+        best_decrease = best_decrease  * coef_reg[best_varID];
+      }
+    }
+    
```

Comment:
Why do we need this?

Reply:
Hey Marvin. That's for keeping the variable importance values on the same scale of a non-regularized model. So if we run a regularized and a non-regularized random forest, their importance values will be similar (and comparable), with the difference that the regularized model will be using fewer variables. I think it's a nice feature to keep, though looking now it seems like could be better written

Reply:
Thanks. Why do you penalize two times? Wouldn't the first if-clause be enough? Then the decrease of Gini impurity would be 

(s1/n1 + s2/n2) * reg - s/n * reg

where _s_ is the squared sum in the current node, _s1_ in the left child node and _s2_ in the right child node, same for the sample sizes _n_, _reg_ is the regularization factor. 

I think currently you are multiplying by the regularization factor again: 

((s1/n1 + s2/n2) * reg - s/n * reg) * reg

Is that on purpose?

Reply:
That's true, the factor is being used two times. I don't recall any reason why that should be used, sorry about that! so the first if-clause should be enough. Do you mind changing this part?

Reply:
No problem, I will change it.

Reply:
Thanks Marvin!